cassandra/policies.py

```diff
@@ -316,6 +327,138 @@ def on_add(self, host):
+        By default, all remote hosts are ignored.
+        """
+        self.local_rack = local_rack
+        self.local_dc = local_dc
+        self.used_hosts_per_remote_dc = used_hosts_per_remote_dc
+        self._live_hosts = {}
+        self._position = 0
+        self._endpoints = []
+        LoadBalancingPolicy.__init__(self)
+
+    def _rack(self, host):
+        return host.rack or self.local_rack
+
+    def _dc(self, host):
+        return host.datacenter or self.local_dc
+
+    def populate(self, cluster, hosts):
+        for (dc, rack), dc_hosts in groupby(hosts, lambda h: (self._dc(h), self._rack(h))):
+            self._live_hosts[(dc, rack)] = tuple(set(dc_hosts))
+
+        if not self.local_rack and not self.local_dc:
+            self._endpoints = [
+                endpoint
+                for endpoint in cluster.endpoints_resolved]
+
+        self._position = randint(0, len(hosts) - 1) if hosts else 0
+
+    def distance(self, host):
+        rack = self._rack(host)
+        dc = self._dc(host)
+        if rack == self.local_rack and dc == self.local_dc:
+            return HostDistance.LOCAL_RACK
+
+        if dc == self.local_dc:
+            return HostDistance.LOCAL
+
+        if not self.used_hosts_per_remote_dc:
+            return HostDistance.IGNORED
+        else:
+            dc_hosts = tuple(host for (datacenter, _), hosts in self._live_hosts.items() if datacenter == dc for host in hosts)
+            if not dc_hosts:
+                return HostDistance.IGNORED
+
+            if host in list(dc_hosts)[:self.used_hosts_per_remote_dc]:
+
+                return HostDistance.REMOTE
+            else:
+                return HostDistance.IGNORED
+
+    def make_query_plan(self, working_keyspace=None, query=None):
+        pos = self._position
+        self._position += 1
+
+        local_rack_live = self._live_hosts.get((self.local_dc, self.local_rack), ())
+        pos = (pos % len(local_rack_live)) if local_rack_live else 0
+        for host in islice(cycle(local_rack_live), pos, pos + len(local_rack_live)):
+            yield host
+
+        local_live = tuple(host for (datacenter, rack), hosts in self._live_hosts.items() if datacenter == self.local_dc and rack != self.local_rack for host in hosts)
+        pos = (pos % len(local_live)) if local_live else 0
+        for host in islice(cycle(local_live), pos, pos + len(local_live)):
+            yield host
+
+        # the dict can change, so get candidate DCs iterating over keys of a copy
+        other_dcs = [dc for (dc, _) in self._live_hosts.copy().keys() if dc != self.local_dc]
+        for dc in other_dcs:
+            remote_live = tuple(host for (datacenter, _), hosts in self._live_hosts.items() if datacenter == dc for host in hosts)
+            for host in remote_live[:self.used_hosts_per_remote_dc]:
+                yield host
+
+    def on_up(self, host):
+        # not worrying about threads because this will happen during
```

Comment:
this kind of comment does make me worry :)

in general sense I think those kind of statements are better to backed up with specific tests (unit, integration) to "prove" them.